pkg/apis/ops/v1alpha1/types.go

```diff
@@ -0,0 +1,191 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type TraceflowPhase string
+
+const (
+	Pending   TraceflowPhase = "Pending"
+	Running   TraceflowPhase = "Running"
+	Succeeded TraceflowPhase = "Succeeded"
+	Failed    TraceflowPhase = "Failed"
+)
+
+type TraceflowComponent string
+
+const (
+	SpoofGuard    TraceflowComponent = "SpoofGuard"
+	LB            TraceflowComponent = "LB"
+	Routing       TraceflowComponent = "Routing"
+	NetworkPolicy TraceflowComponent = "NetworkPolicy"
+	Forwarding    TraceflowComponent = "Forwarding"
+)
+
+type TraceflowAction string
+
+const (
+	Delivered TraceflowAction = "Delivered"
+	Received  TraceflowAction = "Received"
+	Forwarded TraceflowAction = "Forwarded"
+	Dropped   TraceflowAction = "Dropped"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type Traceflow struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   TraceflowSpec   `json:"spec,omitempty"`
+	Status TraceflowStatus `json:"status,omitempty"`
+}
+
+// TraceflowSpec describes the spec of the traceflow.
+type TraceflowSpec struct {
+	Source      Source      `json:"source,omitempty"`
+	Destination Destination `json:"destination,omitempty"`
+	Packet      Packet      `json:"packet,omitempty"`
+}
+
+// Source describes the source spec of the traceflow.
+type Source struct {
+	// Namespace is the source namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the source pod.
+	Pod string `json:"pod,omitempty"`
+}
+
+// Destination describes the destination spec of the traceflow.
+type Destination struct {
+	// Namespace is the destination namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the destination pod, exclusive with destination service.
+	Pod string `json:"pod,omitempty"`
+	// Service is the destination service, exlusive with destination pod.
+	Service string `json:"service,omitempty"`
+	// IP is the destination IP.
+	IP string `json:"IP,omitempty"`
+}
+
+// IPHeader describes spec of an IPv4 header. IPv6 not supported yet.
+type IPHeader struct {
+	// SrcIP is the source IP.
+	SrcIP string `json:"srcIP,omitempty"`
+	// Protocol is the IP protocol.
+	Protocol int32 `json:"protocol,omitempty"`
+	// TTL is the IP TTL.
+	TTL int32 `json:"ttl,omitempty"`
+	// Flags is the flags for IP.
+	Flags int32 `json:"flags,omitempty"`
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	ICMP ICMPEchoRequestHeader `json:"icmp,omitempty"`
+	UDP  UDPHeader             `json:"udp,omitempty"`
+	TCP  TCPHeader             `json:"tcp,omitempty"`
```

Comment:
Why they are restored to value instead of pointer?
To clarify, it's not "all struct fields must be pointer or must not be pointer". A struct field should be a pointer when it's optional, especially the fields are exclusive to each other like I explained before. ICMP, UDP, TCP can never be set together.

Reply:
@tnqn If the type is pointer, we need to verify whether it is nil before use the struct properties, otherwise we'll get nil pointer panic. Is this correct?

Reply:
When using TransportHeader, if you don't check nil, how could you know whether this is a ICMP, UDP or TCP packet? I don't think checking ICMPEchoRequestHeader.ID==0 or UDPHeader.SrcPort==0  is simpler than checking ICMPEchoRequestHeader==nil. And some fields like ID is legal to be set to 0 when it's ICMP request.

Besides, there's nothing wrong to check a field is nil or not when it's optional. It's not different from checking a struct's zero value and more straightforward.

Reply:
@gran-vmv Thought over, I think Quan is right. Pointer helps tell if these headers are unset.

Reply:
Please update codegen for this.

Reply:
Updated.